File: backend/src/routes/message.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from datetime import datetime
from typing import Optional

from ..config.database import get_db
from ..utils import success_response, error_response, get_current_active_user
from ..models import User, Message

router = APIRouter(prefix="/message", tags=["消息"])


@router.get("/categories")
async def get_message_categories(
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    categories = [
        {"key": "system", "name": "系统通知", "icon": "bell"},
        {"key": "fresh", "name": "新鲜事", "icon": "fire"},
        {"key": "announcement", "name": "公告", "icon": "notification"},
        {"key": "activity", "name": "活动中心", "icon": "gift"},
        {"key": "interaction", "name": "互动消息", "icon": "chat"}
    ]
    
    result = await db.execute(
        select(Message.category, Message.id).where(
            Message.user_id == current_user.id,
            Message.is_read == False
        )
    )
    unread_messages = result.all()
    
    unread_count = {}
    for category, _ in unread_messages:
        unread_count[category] = unread_count.get(category, 0) + 1
    
    for cat in categories:
        cat["unread_count"] = unread_count.get(cat["key"], 0)
    
    return success_response(categories)


@router.get("/list")
async def get_messages(
    category: Optional[str] = None,
    page: int = 1,
    page_size: int = 20,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    try:
        query = select(Message).where(Message.user_id == current_user.id)
        
        if category:
            query = query.where(Message.category == category)
        
        query = query.order_by(Message.created_at.desc())
        
        total_result = await db.execute(select(Message.id).where(Message.user_id == current_user.id))
        total = len(total_result.all())
        
        result = await db.execute(
            query.offset((page - 1) * page_size).limit(page_size)
        )
        messages = result.scalars().all()
        
        return success_response({
            "list": [{
                "id": msg.id,
                "category": msg.category,
                "title": msg.title,
                "content": msg.content,
                "is_read": msg.is_read,
                "created_at": msg.created_at.isoformat()
            } for msg in messages],
            "total": total,
            "page": page,
            "page_size": page_size
        })
    
    except Exception as e:
        return error_response(f"获取消息列表失败: {str(e)}")


@router.post("/{message_id}/read")
async def mark_message_read(
    message_id: int,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    try:
        result = await db.execute(
            select(Message).where(
                Message.id == message_id,
                Message.user_id == current_user.id
            )
        )
        message = result.scalar_one_or_none()
        
        if not message:
            return error_response("消息不存在")
        
        message.is_read = True
        message.read_at = datetime.utcnow()
        await db.commit()
        
        return success_response(None, "标记已读成功")
    
    except Exception as e:
        await db.rollback()
        return error_response(f"标记已读失败: {str(e)}")


@router.post("/read-all")
async def mark_all_messages_read(
    category: Optional[str] = None,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    try:
        query = select(Message).where(
            Message.user_id == current_user.id,
            Message.is_read == False
        )
        
        if category:
            query = query.where(Message.category == category)
        
        result = await db.execute(query)
        messages = result.scalars().all()
        
        for msg in messages:
            msg.is_read = True
            msg.read_at = datetime.utcnow()
        
        await db.commit()
        
        return success_response({"count": len(messages)}, "全部标记已读成功")
    
    except Exception as e:
        await db.rollback()
        return error_response(f"标记已读失败: {str(e)}")


@router.delete("/{message_id}")
async def delete_message(
    message_id: int,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    try:
        result = await db.execute(
            select(Message).where(
                Message.id == message_id,
                Message.user_id == current_user.id
            )
        )
        message = result.scalar_one_or_none()
        
        if not message:
            return error_response("消息不存在")
        
        await db.delete(message)
        await db.commit()
        
        return success_response(None, "删除成功")
    
    except Exception as e:
        await db.rollback()
        return error_response(f"删除失败: {str(e)}")

File: backend/src/routes/message.ts
import express, { Request, Response } from 'express';
import { getDb } from '../database';
import { authenticateToken } from '../middleware/auth';
import { successResponse, errorResponse, paginatedResponse } from '../utils/response';
import { Message } from '../types';

const router = express.Router();

router.use(authenticateToken);

interface AuthRequest extends Request {
  user?: { userId: string; idCard: string };
}

router.get('/', (req: AuthRequest, res: Response) => {
  try {
    const db = getDb();
    const userId = req.user?.userId;
    const { type, is_read, page = 1, pageSize = 20 } = req.query;

    let whereClause = 'WHERE user_id = ?';
    const params: any[] = [userId];

    if (type && type !== 'all') {
      whereClause += ' AND type = ?';
      params.push(type);
    }

    if (is_read !== undefined && is_read !== '') {
      whereClause += ' AND is_read = ?';
      params.push(is_read === 'true' ? 1 : 0);
    }

    const countStmt = db.prepare(`SELECT COUNT(*) as total FROM messages ${whereClause}`);
    const { total } = countStmt.get(...params) as any;

    const offset = (Number(page) - 1) * Number(pageSize);
    const stmt = db.prepare(`
      SELECT * FROM messages ${whereClause}
      ORDER BY created_at DESC
      LIMIT ? OFFSET ?
    `);
    params.push(Number(pageSize), offset);

    const messages = stmt.all(...params) as any[];

    return paginatedResponse(res, messages, total, Number(page), Number(pageSize));
  } catch (err) {
    return errorResponse(res, '获取消息列表失败', 500);
  }
});

router.get('/unread-count', (req: AuthRequest, res: Response) => {
  try {
    const db = getDb();
    const userId = req.user?.userId;

    const counts = db.prepare(`
      SELECT 
        type,
        COUNT(*) as count
      FROM messages 
      WHERE user_id = ? AND is_read = 0
      GROUP BY type
    `).all(userId) as any[];

    const totalUnread = counts.reduce((sum, item: any) => sum + item.count, 0);

    const typeMap: Record<string, number> = {
      system: 0,
      business: 0,
      approval: 0,
      service: 0,
    };

    counts.forEach((item: any) => {
      typeMap[item.type] = item.count;
    });

    return successResponse(res, {
      total: totalUnread,
      byType: typeMap,
    });
  } catch (err) {
    return errorResponse(res, '获取未读消息数失败', 500);
  }
});

router.get('/:id', (req: AuthRequest, res: Response) => {
  try {
    const db = getDb();
    const userId = req.user?.userId;
    const { id } = req.params;

    const message = db.prepare(
      'SELECT * FROM messages WHERE id = ? AND user_id = ?'
    ).get(id, userId) as any;

    if (!message) {
      return errorResponse(res, '消息不存在', 404);
    }

    if (!message.is_read) {
      db.prepare(
        'UPDATE messages SET is_read = 1, read_at = ? WHERE id = ?'
      ).run(new Date().toISOString(), id);
    }

    return successResponse(res, message);
  } catch (err) {
    return errorResponse(res, '获取消息详情失败', 500);
  }
});

router.post('/:id/read', (req: AuthRequest, res: Response) => {
  try {
    const db = getDb();
    const userId = req.user?.userId;
    const { id } = req.params;

    const message = db.prepare(
      'SELECT * FROM messages WHERE id = ? AND user_id = ?'
    ).get(id, userId) as any;

    if (!message) {
      return errorResponse(res, '消息不存在', 404);
    }

    db.prepare(
      'UPDATE messages SET is_read = 1, read_at = ? WHERE id = ?'
    ).run(new Date().toISOString(), id);

    return successResponse(res, null, '标记已读成功');
  } catch (err) {
    return errorResponse(res, '标记已读失败', 500);
  }
});

router.post('/read-all', (req: AuthRequest, res: Response) => {
  try {
    const db = getDb();
    const userId = req.user?.userId;
    const { type } = req.body;

    let whereClause = 'user_id = ? AND is_read = 0';
    const params: any[] = [userId];

    if (type && type !== 'all') {
      whereClause += ' AND type = ?';
      params.push(type);
    }

    const result = db.prepare(`
      UPDATE messages 
      SET is_read = 1, read_at = ?
      WHERE ${whereClause}
    `).run(new Date().toISOString(), ...params);

    return successResponse(res, {
      updated: result.changes,
    }, `已标记${result.changes}条消息为已读`);
  } catch (err) {
    return errorResponse(res, '批量标记已读失败', 500);
  }
});

router.post('/:id/delete', (req: AuthRequest, res: Response) => {
  try {
    const db = getDb();
    const userId = req.user?.userId;
    const { id } = req.params;

    const message = db.prepare(
      'SELECT * FROM messages WHERE id = ? AND user_id = ?'
    ).get(id, userId) as any;

    if (!message) {
      return errorResponse(res, '消息不存在', 404);
    }

    db.prepare('DELETE FROM messages WHERE id = ?').run(id);

    return successResponse(res, null, '删除成功');
  } catch (err) {
    return errorResponse(res, '删除失败', 500);
  }
});

router.post('/delete-read', (req: AuthRequest, res: Response) => {
  try {
    const db = getDb();
    const userId = req.user?.userId;

    const result = db.prepare(`
      DELETE FROM messages 
      WHERE user_id = ? AND is_read = 1
    `).run(userId);

    return successResponse(res, {
      deleted: result.changes,
    }, `已删除${result.changes}条已读消息`);
  } catch (err) {
    return errorResponse(res, '清空已读消息失败', 500);
  }
});

router.get('/settings', (req: AuthRequest, res: Response) => {
  try {
    const db = getDb();
    const userId = req.user?.userId;

    const settings = db.prepare(
      'SELECT * FROM notification_settings WHERE user_id = ?'
    ).get(userId) as any;

    if (settings) {
      return successResponse(res, settings);
    }

    return successResponse(res, {
      push_enabled: true,
      system_enabled: true,
      business_enabled: true,
      approval_enabled: true,
      service_enabled: true,
      sound_enabled: true,
      vibration_enabled: false,
      quiet_start: '22:00',
      quiet_end: '07:00',
      quiet_enabled: false,
    });
  } catch (err) {
    return errorResponse(res, '获取通知设置失败', 500);
  }
});

router.post('/settings', (req: AuthRequest, res: Response) => {
  try {
    const db = getDb();
    const userId = req.user?.userId;
    const {
      push_enabled,
      system_enabled,
      business_enabled,
      approval_enabled,
      service_enabled,
      sound_enabled,
      vibration_enabled,
      quiet_start,
      quiet_end,
      quiet_enabled,
    } = req.body;

    const existing = db.prepare(
      'SELECT id FROM notification_settings WHERE user_id = ?'
    ).get(userId) as any;

    const now = new Date().toISOString();

    if (existing) {
      db.prepare(`
        UPDATE notification_settings SET
          push_enabled = ?,
          system_enabled = ?,
          business_enabled = ?,
          approval_enabled = ?,
          service_enabled = ?,
          sound_enabled = ?,
          vibration_enabled = ?,
          quiet_start = ?,
          quiet_end = ?,
          quiet_enabled = ?,
          updated_at = ?
        WHERE user_id = ?
      `).run(
        push_enabled ? 1 : 0,
        system_enabled ? 1 : 0,
        business_enabled ? 1 : 0,
        approval_enabled ? 1 : 0,
        service_enabled ? 1 : 0,
        sound_enabled ? 1 : 0,
        vibration_enabled ? 1 : 0,
        quiet_start || '22:00',
        quiet_end || '07:00',
        quiet_enabled ? 1 : 0,
        now,
        userId
      );
    } else {
      db.prepare(`
        INSERT INTO notification_settings (
          user_id, push_enabled, system_enabled, business_enabled,
          approval_enabled, service_enabled, sound_enabled, vibration_enabled,
          quiet_start, quiet_end, quiet_enabled, created_at, updated_at
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
      `).run(
        userId,
        push_enabled ? 1 : 0,
        system_enabled ? 1 : 0,
        business_enabled ? 1 : 0,
        approval_enabled ? 1 : 0,
        service_enabled ? 1 : 0,
        sound_enabled ? 1 : 0,
        vibration_enabled ? 1 : 0,
        quiet_start || '22:00',
        quiet_end || '07:00',
        quiet_enabled ? 1 : 0,
        now,
        now
      );
    }

    return successResponse(res, null, '设置保存成功');
  } catch (err) {
    return errorResponse(res, '保存设置失败', 500);
  }
});

export default router;
